src/components/include/utils/timer_thread.h

```diff
@@ -0,0 +1,421 @@
+    // Sleep
+    int32_t wait_milliseconds_left = TimerDelegate::calculateMillisecondsLeft();
+    ConditionalVariable::WaitStatus wait_status =
+        termination_condition_.WaitFor(auto_lock, wait_milliseconds_left);
+    // Quit sleeping or continue sleeping in case of spurious wake up
+    if (ConditionalVariable::kTimeout == wait_status ||
+        wait_milliseconds_left <= 0) {
+      LOG4CXX_TRACE(logger_,
+                    "Timer timeout " << wait_milliseconds_left << " ms");
+      timer_thread_->onTimeOut();
+      return;
+    } else {
+      LOG4CXX_DEBUG(logger_,
+                    "Timeout reset force: " << TimerDelegate::timeout_seconds_);
+      return;
+    }
+  }
+}
+
+template <class T>
+void TimerThread<T>::TimerLooperDelegate::threadMain() {
+  using sync_primitives::ConditionalVariable;
+  sync_primitives::AutoLock auto_lock(TimerDelegate::state_lock_);
+  TimerDelegate::stop_flag_ = false;
+  while (!TimerDelegate::stop_flag_) {
+    int32_t wait_milliseconds_left = TimerDelegate::calculateMillisecondsLeft();
+    ConditionalVariable::WaitStatus wait_status =
+        TimerDelegate::termination_condition_.WaitFor(auto_lock,
+                                                      wait_milliseconds_left);
+    // Quit sleeping or continue sleeping in case of spurious wake up
+    if (ConditionalVariable::kTimeout == wait_status ||
+        wait_milliseconds_left <= 0) {
+      LOG4CXX_TRACE(logger_,
+                    "Timer timeout " << wait_milliseconds_left << " ms");
+      TimerDelegate::timer_thread_->onTimeOut();
+    } else {
+      LOG4CXX_DEBUG(logger_,
+                    "Timeout reset force: " << TimerDelegate::timeout_seconds_);
+    }
+  }
+}
+
+template <class T>
+void TimerThread<T>::TimerDelegate::exitThreadMain() {
+  sync_primitives::AutoLock auto_lock(state_lock_);
+  stop_flag_ = true;
+  termination_condition_.NotifyOne();
+}
+
+template <class T>
+void TimerThread<T>::TimerDelegate::setTimeOut(const uint32_t timeout_seconds) {
+  timeout_seconds_ = timeout_seconds;
+  termination_condition_.NotifyOne();
+}
+
+template <class T>
+int32_t
+TimerThread<T>::TimerThread::TimerDelegate::calculateMillisecondsLeft() {
+  time_t cur_time = time(NULL);
+  time_t end_time = std::numeric_limits<time_t>::max();
+  if (TimerDelegate::timeout_seconds_ + cur_time >
+      TimerDelegate::timeout_seconds_) {  // no overflow occurred
+    end_time = cur_time + TimerDelegate::timeout_seconds_;
+  }
+
+  int64_t wait_seconds_left =
+      static_cast<int64_t>(difftime(end_time, cur_time));
+  int32_t wait_milliseconds_left = std::numeric_limits<int32_t>::max();
+  const int32_t milliseconds_in_second = 1000;
+  if (wait_seconds_left <
+      std::numeric_limits<int32_t>::max() / milliseconds_in_second) {
+    wait_milliseconds_left = milliseconds_in_second * wait_seconds_left;
+  }
+  return wait_milliseconds_left;
+}
+
+}  // namespace timer
+
+#endif  // SRC_COMPONENTS_INCLUDE_UTILS_TIMER_THREAD_H_
```

Comment:
@LitvinenkoIra please remove this file and try to check build without it. It looks like redundant